internal/server/handler.go

```diff
@@ -37,15 +39,45 @@ func (s *Server) handler() http.HandlerFunc {
 			installationID = s[1]
 		}
 
-		r.URL.Scheme = s.cfg.S3Settings.Scheme
-		r.URL.Host = host
+		// Strip the bucket name from the path which gets added by Minio
+		// if the S3 hostname does not match a URL pattern.
+		objectName := strings.TrimPrefix(r.URL.Path, "/"+s.cfg.S3Settings.Bucket)
+
+		// Rebuild the URL from scratch, using s3utils.EncodePath on the unescaped
+		// objectName from the path.
+		//
+		// When Mattermost makes an S3 request, the minio library already calls
+		// s3utils.EncodePath, translating (among other characters) both ' ' to %20 and
+		// '+' to %2B. This is actually more strict than RFC 3986 requires, since a
+		// '+' in the path doesn't actually need to be escaped.
+		//
+		// When Bifrost receives the request, net.URL happily unescapes both characters.
```

Comment:
So IIUC, there are two phases of URL encoding happening. One just for sending the URL to AWS, and another to calculate the signature. If you directly use Minio -> AWS, both use `s3utils.EncodePath`, but if you use Minio -> Bifros-> AWS, then the URL uses normal URL encoding, but the signature computation uses `s3utils.EncodePath`. And hence the mismatch. Am I right?

Reply:
Yes, but to clarify the path from `Minio -> AWS` and `Minio -> Bifrost` yields identical request payloads from MM. Where things went wrong is `Bifrost -> AWS` which didn't use Minio but just a basic HTTP request, and thus incurred the normal URL encoding to which you refer.

Reply:
Got it.